src/core/harpoon.ml

```diff
@@ -304,14 +304,130 @@ module Tactic (T : TacticContext) = struct
       ; solution = None
       }
     in
-    T.remove_current_subgoal ();
-    T.add_subgoal new_state;
+    tctx.remove_current_subgoal ();
+    tctx.add_subgoal new_state;
     prepend_commands
       [ Comp.IH (m, name) ]
       (Comp.incomplete_proof new_state)
     |> solve' g
 end
 
+module Automation = struct
+  open Comp
+
+  type t = unit proof_state -> Tactic.tactic_context -> bool
+
+  let auto_intros : t =
+    fun g tctx ->
+    let (t, _) = g.goal in
+    dprintf
+      (fun p ->
+        p.fmt
+          "[auto_intros]: invoked on %a"
+          (P.fmt_ppr_cmp_typ g.context.cD Pretty.std_lvl) t
+      );
+    match t with
+    | TypArr _ | TypPiBox _ ->
+       Tactic.intros None g tctx;
+       true
+    | _ -> false
+
+  (** Solve {v ... -> P -> ... -> P v} case automatically.
+      For example,
+      this function will resolve
+
+        x: [|- a]
+      -------------
+      [|- a]
+   *)
+  let auto_solve_trivial : t =
+    fun g tctx ->
+    let { cD; cG; _ } = g.context in
+    let m_is_witness (m : LF.ctyp_decl) =
+      dprintf
+        (fun p ->
+          p.fmt
+            "@[<v>[auto_solve_trivial] witness candidate = %a@]"
+            (P.fmt_ppr_lf_ctyp_decl cD Pretty.std_lvl) m
+        );
+      match m with
+      | LF.Decl (_, mtyp, _) ->
+         Whnf.convCTyp g.goal (Comp.TypBox (Loc.ghost, mtyp), LF.MShift 0)
```

Comment:
Good observation that `MShift 0` is the identity substitution; we have a definition in `whnf.ml` that defines `m_id = MShift 0`; you may want to use that instead of explicitly forming the identity substitution here.